v2/coordinator-specification.md

```diff
@@ -0,0 +1,691 @@
+
+## Configuration
+
+### SELECTIVE_DELAY_MS
+
+The delay in milliseconds between the receipt of fill requests and fill approvals (default: 1000ms).
+
+### EXPIRATION_DURATION_SECONDS
+
+The amount of seconds an approval is valid for (default: 60 seconds). This parameter may be tweaked based off of the amount of congestion on the Ethereum network or the desired UX of your application. Users who are manually submitting transaction through a UI may need longer expiration times than a programmatic trader.
+
+## State
+
+The coordinator server must maintain state in order to determine the validity of transaction requests.
+
+- The server must be able to identify orders for which it has issued a cancellation receipt (e.g by storing their hashes). This information can be discarded after the order has expired or been invalidated.
+- For each order with at least one approved fill, the server must store the sum of approved `takerAssetFillAmount`s for each approved taker. This information can be discarded after the order has been filled or invalidated. Note that this information should persist even if outstanding approvals that involve the order have expired.
+- For each order with at least one approved fill, the server must store any outstanding approval signatures for fill transactions that contain the order, as well as the corresponding expiration timestamp of each approval signature.
+- The server must be able to identify transactions that it has already approved (e.g by storing its hash). This information can be discarded after the transaction has been executed, invalidated, or if the approval has expired.
+
+## Handling fills
+
+Fill transaction requests should be rejected under the following conditions:
+
+- The transaction is in any way invalid (incorrect formatting, signature, unsupported function, etc).
+- A transaction with the same hash has already been approved.
+- The taker has requested fills for an included order with total `takerAssetFillAmount`s that exceed the `takerAssetAmount` of the order. This creates a cost for requesting fills and should prevent takers from frequently requesting fills that they do not intend on executing.
+- An included order has been soft cancelled (TODO: Error response should include cancelled orders)
+
+All other fill requests must be approved by the coordinator server exactly `SELECTIVE_DELAY_MS` after the request is received (TODO: how much on-chain validation should be done?).
```

Comment:
Note on this TODO: I feel like there should actually be pretty minimal on-chain validation. If too much is done, I can see a lot of edge cases (for example, order is cancelled on-chain but the tx is for `fillOrderNoThrow`, so technically the tx is still valid).